lib/bloom/sbbf.ts

```diff
@@ -0,0 +1,396 @@
+                }
+            }
+        }
+    }
+
+    /**
+     * @function blockCheck
+     * @description check if a hashValue exists for this filter
+     * @param b: the block to check for inclusion
+     * @param hashValue: the hash to check for  should be long
+     * @private
+     *
+     * @return true if hashed item is __probably__ in the data set represented by this filter
+     * @return false if it is __definitely not__ in the data set.
+     */
+    static blockCheck(b: Block, hashValue: Long): boolean {
+        const masked: Block = this.mask(hashValue)
+        for (let i = 0; i < masked.length; i++) {
+            for (let j = 0; j < this.WORD_SIZE; j++) {
+                const isSet = masked[i] & (2 ** j)
+                if (isSet) {
+                    const match = b[i] & (2 ** j)
+                    if (!match) {
+                        return false
+                    }
+                }
+            }
+        }
+        return true
+    }
+
+    /**
+     * Instance
+     */
+
+    private splitBlockFilter: Array<Block> = []
+    private desiredFalsePositiveRate: number = SplitBlockBloomFilter.DEFAULT_FALSE_POSITIVE_RATE
+    private numBlocks: number = 0
+    private numDistinctValues: number = SplitBlockBloomFilter.DEFAULT_DISTINCT_VALUES
+    private hashStrategy = new parquet_thrift.BloomFilterHash(parquet_thrift.XxHash)
+
+    private isInitialized(): boolean { return this.splitBlockFilter.length > 0 }
+
+    getFalsePositiveRate(): number { return this.desiredFalsePositiveRate }
+    getNumDistinct(): number { return this.numDistinctValues }
+    getNumFilterBlocks(): number { return this.splitBlockFilter.length }
+    getFilter(): Array<Block> { return this.splitBlockFilter }
+
+    /**
+     * @function  optNumFilterBytes
+     * @description return the actual number of filter bytes set; if the option to numBytes
+     *     was called, this value will be returned. If the options for preferred FPR
+     *     and/or numDistinct were called, this function returns the calculated value.
+     */
+    getNumFilterBytes(): number {
+        return this.numBlocks * SplitBlockBloomFilter.BITS_PER_BLOCK >>> 3
+    }
+
+    /**
+     * @function setOptionFalsePositiveRate
+     * @description set the desired false positive percentage for this Bloom filter.
+     *     defaults to SplitBlockBLoomFilter.DEFAULT_FALSE_POSITIVE_RATE
+     *     This function does nothing if the filter has already been allocated.
+     * @param proportion: number, between 0.0 and 1.0, exclusive
+     */
+    setOptionFalsePositiveRate(proportion: number): SplitBlockBloomFilter {
+        if (this.isInitialized()) {
+            console.error("filter already initialized. options may no longer be changed.")
+            return this
+        }
+        if (proportion <= 0.0 || proportion >= 1.0) {
+            console.error("falsePositiveProbability. Must be < 1.0 and > 0.0")
+            return this
```

Comment:
There are a few of these swallowed errors. Would we rather throw?

Reply:
I don't know. I was thinking about whether you'd rather it stop the app altogether or you just want to know that it's not going to do the thing you think it is.  Trying to set a value out of range would leave it as a default and it would still work.  

I definitely don't want to throw if the case where init()  is called more than once, or when caller tries to set options after initializing; I'd rather just let the caller know they're doing it in the logs. If you have some good reasons to throw anyway I'd love to know.

Reply:
Hmm.... I agree with the double init case (effectively it just makes init idempotent, which is good).

The post init case also is ok at least in the case where `this.desiredFalsePositiveRate === proportion`.

I'm less sure about the post init case where a change is attempted.

Finally this case. This is an error that is hard to catch, but always wrapping these in a try catch is also not very nice.

All that to say... Let's leave it.